### PR TITLE
Feature - Multiple Access Control Permissions

### PR DESCRIPTION
**Pull request by harryMartin678**
Backbase Email: harry@backbase.com

Some of digital assists endpoints are protected with multiple function group permissions which we want to represent elegantly. This PR keeps the old properties for access control i.e.  x-BbAccessControl-resource, etc... but adds an optional x-BbAccessControls with the format below:
```
x-BbAccessControls:
  description: Description including references to the permissions (a), (b) etc...
  permissions:
    - resource: Permission_a_Resource
      function: Permission_a_Function
      privilege: Permission_a_Privilege
    - resource: Permission_b_Resource
       function: Permission_b_Function
       privilege: Permission_b_Privilege
```

Theses permissions appears in the generated docs side by side. 

Example: 
Screenshot 2024-04-19 at 15 36 31

### CODE REVIEW

File: boat-scaffold/src/main/templates/boat-docs/styles.mustache [hunk 1]
```diff
@@ -150,7 +165,7 @@ div.secured-by > pre:nth-child(2) {
   padding-top: 8px;
 }
 
-div.secured-by > pre:last-child {
+div.secured-by,div.secured-by-multiple > pre:last-child {
```
Comment (daiscog) [inline boat-scaffold/src/main/templates/boat-docs/styles.mustache:168]: Should this actually be as follows:

```suggestion
div.secured-by > pre:last-child, div.secured-by-multiple > pre:last-child {
```

File: boat-scaffold/src/main/templates/boat-docs/styles.mustache [hunk 2]
```diff
@@ -136,7 +145,13 @@ div.secured-by > h2:only-child {
...
 }
 
-div.secured-by > pre {
+div.secured-by-multiple > pre:nth-child(1) {
+  border-top-left-radius: 6px;
+  border-top-right-radius: 6px;
+  padding-top: 8px;
+}
+
+div.secured-by,div.secured-by-multiple > pre {
```
Comment (daiscog) [inline boat-scaffold/src/main/templates/boat-docs/styles.mustache:154]:
```suggestion
div.secured-by > pre, div.secured-by-multiple > pre {
```

**Comment by jjjasper** [comment]
What is the spec supposed to look like to generate these docs and how will that express the difference between "in addition to" and "one of"?

There is a linting rule that enforces x-bb-access-control tags in specs.

There is a guideline that says to "use plain english when things are more complicated"

**Comment by harryMartin678** [comment]
> Details

Currently the format is as below: 

Spec:
`x-BbAccessControls:
        - resource: <RESOURCE_1>
          function: <FUNCTION_1>
          privilege: <PRIVILEGE_1>
        - resource: <RESOURCE_2>
           function: <FUNCTION_2>
           privilege: <PRIVILEGE_2>`

Example:
`x-BbAccessControls:
        - resource: WALLET
          function: MANAGE_PAYMENT_CARDS
          privilege: READ_PAYMENT_CARD
        - resource: CARDS
          function: MANAGE
          privilege: READ_CARD`

It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper

**Comment by jjjasper** [comment]
> [> Details](https://github.com/Backbase/backbase-openapi-tools/pull/799#issuecomment-2045312878)
> 
> An example of the format is here: https://github.com/Backbase/backbase-openapi-tools/pull/799/files#diff-5b822b71b3ff4e24208c226f50c5a6041db8969ff8aeb4b26649052197632ef7R215
> 
> It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper

My concern is that it will become very complex to support all different permutations of how access control is enforced. It may be possible to reflect _Function A create OR Function B create AND one of Function C view, edit, delete, approve OR Function D view_ but it will still need text, for instance because "Function D" is about "managing confidential batches", or, as often is the case: the business function name is subject to the configuration of a service.

**Comment by harryMartin678** [comment]
> > [> Details](https://github.com/Backbase/backbase-openapi-tools/pull/799#issuecomment-2045312878)
> > An example of the format is here: https://github.com/Backbase/backbase-openapi-tools/pull/799/files#diff-5b822b71b3ff4e24208c226f50c5a6041db8969ff8aeb4b26649052197632ef7R215
> > It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper
> 
> My concern is that it will become very complex to support all different permutations of how access control is enforced. It may be possible to reflect _Function A create OR Function B create AND one of Function C view, edit, delete, approve OR Function D view_ but it will still need text, for instance because "Function D" is about "managing confidential batches", or, as often is the case: the business function name is subject to the configuration of a service.

What about the following format? 

      x-BbAccessControls:
        description: Require (3) with either one of (1) or (2)
        permissions:
          - resource: Personal Finance Management
            function: Manage Cards
            privilege: create
          - resource: Support Access
            function: Cards - Order new card
            privilege: execute
          - resource: Employee
            function: Act on behalf of
            privilege: execute

With the relationship between the permission described by a text rather than some sort of AND/OR mechanic to keep it simpler. The business function name in the current BOAT single permission has a hardcoded value (regardless of configuration) so multiple permissions would be the same. @jjjasper

**Comment by jjjasper** [comment]
> > > [> Details](https://github.com/Backbase/backbase-openapi-tools/pull/799#issuecomment-2045312878)
> > > An example of the format is here: https://github.com/Backbase/backbase-openapi-tools/pull/799/files#diff-5b822b71b3ff4e24208c226f50c5a6041db8969ff8aeb4b26649052197632ef7R215
> > > It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper
> > 
> > 
> > My concern is that it will become very complex to support all different permutations of how access control is enforced. It may be possible to reflect _Function A create OR Function B create AND one of Function C view, edit, delete, approve OR Function D view_ but it will still need text, for instance because "Function D" is about "managing confidential batches", or, as often is the case: the business function name is subject to the configuration of a service.
> 
> What about the following format?
> 
> ```
>   x-BbAccessControls:
>     description: Require (3) with either one of (1) or (2)
>     permissions:
>       - resource: Personal Finance Management
>         function: Manage Cards
>         privilege: create
>       - resource: Support Access
>         function: Cards - Order new card
>         privilege: execute
>       - resource: Employee
>         function: Act on behalf of
>         privilege: execute
> ```
> 
> With the relationship between the permission described by a text rather than some sort of AND/OR mechanic to keep it simpler. The business function name in the current BOAT single permission has a hardcoded value (regardless of configuration) so multiple permissions would be the same. @jjjasper

How could that be used for something like this found in the wild
```
**Resource:** Payments;Product Summary (for enrichment)
**Function:** Different functions per payment type, eg 'SEPA CT', 'US Domestic Wire', etc; Product Summary
**Privilege:** create on the payments function;view on Product Summary
```

**Comment by harryMartin678** [comment]
> > > > [> Details](https://github.com/Backbase/backbase-openapi-tools/pull/799#issuecomment-2045312878)
> > > > An example of the format is here: https://github.com/Backbase/backbase-openapi-tools/pull/799/files#diff-5b822b71b3ff4e24208c226f50c5a6041db8969ff8aeb4b26649052197632ef7R215
> > > > It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper
> > > 
> > > 
> > > My concern is that it will become very complex to support all different permutations of how access control is enforced. It may be possible to reflect _Function A create OR Function B create AND one of Function C view, edit, delete, approve OR Function D view_ but it will still need text, for instance because "Function D" is about "managing confidential batches", or, as often is the case: the business function name is subject to the configuration of a service.
> > 
> > 
> > What about the following format?
> > ```
> >   x-BbAccessControls:
> >     description: Require (3) with either one of (1) or (2)
> >     permissions:
> >       - resource: Personal Finance Management
> >         function: Manage Cards
> >         privilege: create
> >       - resource: Support Access
> >         function: Cards - Order new card
> >         privilege: execute
> >       - resource: Employee
> >         function: Act on behalf of
> >         privilege: execute
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > With the relationship between the permission described by a text rather than some sort of AND/OR mechanic to keep it simpler. The business function name in the current BOAT single permission has a hardcoded value (regardless of configuration) so multiple permissions would be the same. @jjjasper
> 
> How could that be used for something like this found in the wild
> 
> ```
> **Resource:** Payments;Product Summary (for enrichment)
> **Function:** Different functions per payment type, eg 'SEPA CT', 'US Domestic Wire', etc; Product Summary
> **Privilege:** create on the payments function;view on Product Summary
> ```
So I don't know if I fully understand the example or if I have included every case in the yaml below but this how the format it could be used for that example. 

  x-BbAccessControls:
    description: (1) required for SEPA CT payment, (2) for US domestic Wire and (3) for enrichment. 
    permissions:
      - resource: Payments
        function: SEPA CT
        privilege: create
      - resource: Payments
        function: US Domestic Wire
        privilege: create
      - resource: Product Summary
        function: Enrichment
        privilege: view

@jjjasper

**Comment by jjjasper** [comment]
> > > > > [> Details](https://github.com/Backbase/backbase-openapi-tools/pull/799#issuecomment-2045312878)
> > > > > An example of the format is here: https://github.com/Backbase/backbase-openapi-tools/pull/799/files#diff-5b822b71b3ff4e24208c226f50c5a6041db8969ff8aeb4b26649052197632ef7R215
> > > > > It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper
> > > > 
> > > > 
> > > > My concern is that it will become very complex to support all different permutations of how access control is enforced. It may be possible to reflect _Function A create OR Function B create AND one of Function C view, edit, delete, approve OR Function D view_ but it will still need text, for instance because "Function D" is about "managing confidential batches", or, as often is the case: the business function name is subject to the configuration of a service.
> > > 
> > > 
> > > What about the following format?
> > > ```
> > >   x-BbAccessControls:
> > >     description: Require (3) with either one of (1) or (2)
> > >     permissions:
> > >       - resource: Personal Finance Management
> > >         function: Manage Cards
> > >         privilege: create
> > >       - resource: Support Access
> > >         function: Cards - Order new card
> > >         privilege: execute
> > >       - resource: Employee
> > >         function: Act on behalf of
> > >         privilege: execute
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > With the relationship between the permission described by a text rather than some sort of AND/OR mechanic to keep it simpler. The business function name in the current BOAT single permission has a hardcoded value (regardless of configuration) so multiple permissions would be the same. @jjjasper
> > 
> > 
> > How could that be used for something like this found in the wild
> > ```
> > **Resource:** Payments;Product Summary (for enrichment)
> > **Function:** Different functions per payment type, eg 'SEPA CT', 'US Domestic Wire', etc; Product Summary
> > **Privilege:** create on the payments function;view on Product Summary
> > ```
> 
> So I don't know if I fully understand the example or if I have included every case in the yaml below but this how the format it could be used for that example.
> 
> ```
> x-BbAccessControls:
>   description: (1) required for SEPA CT payment, (2) for US domestic Wire and (3) for enrichment. 
>   permissions:
>     - resource: Payments
>       function: SEPA CT
>       privilege: create
>     - resource: Payments
>       function: US Domestic Wire
>       privilege: create
>     - resource: Product Summary
>       function: Enrichment
>       privilege: view
> ```
> 
> @jjjasper

The problem in this case, is that which permission is required for each payment type is configured in the service. It is not defined on spec level.

**Comment by harryMartin678** [comment]
> > > > > > [> Details](https://github.com/Backbase/backbase-openapi-tools/pull/799#issuecomment-2045312878)
> > > > > > An example of the format is here: https://github.com/Backbase/backbase-openapi-tools/pull/799/files#diff-5b822b71b3ff4e24208c226f50c5a6041db8969ff8aeb4b26649052197632ef7R215
> > > > > > It doesn't have any information about how the permissions relate to each other. But it could be possible to add this. @jjjasper
> > > > > 
> > > > > 
> > > > > My concern is that it will become very complex to support all different permutations of how access control is enforced. It may be possible to reflect _Function A create OR Function B create AND one of Function C view, edit, delete, approve OR Function D view_ but it will still need text, for instance because "Function D" is about "managing confidential batches", or, as often is the case: the business function name is subject to the configuration of a service.
> > > > 
> > > > 
> > > > What about the following format?
> > > > ```
> > > >   x-BbAccessControls:
> > > >     description: Require (3) with either one of (1) or (2)
> > > >     permissions:
> > > >       - resource: Personal Finance Management
> > > >         function: Manage Cards
> > > >         privilege: create
> > > >       - resource: Support Access
> > > >         function: Cards - Order new card
> > > >         privilege: execute
> > > >       - resource: Employee
> > > >         function: Act on behalf of
> > > >         privilege: execute
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > With the relationship between the permission described by a text rather than some sort of AND/OR mechanic to keep it simpler. The business function name in the current BOAT single permission has a hardcoded value (regardless of configuration) so multiple permissions would be the same. @jjjasper
> > > 
> > > 
> > > How could that be used for something like this found in the wild
> > > ```
> > > **Resource:** Payments;Product Summary (for enrichment)
> > > **Function:** Different functions per payment type, eg 'SEPA CT', 'US Domestic Wire', etc; Product Summary
> > > **Privilege:** create on the payments function;view on Product Summary
> > > ```
> > 
> > 
> > So I don't know if I fully understand the example or if I have included every case in the yaml below but this how the format it could be used for that example.
> > ```
> > x-BbAccessControls:
> >   description: (1) required for SEPA CT payment, (2) for US domestic Wire and (3) for enrichment. 
> >   permissions:
> >     - resource: Payments
> >       function: SEPA CT
> >       privilege: create
> >     - resource: Payments
> >       function: US Domestic Wire
> >       privilege: create
> >     - resource: Product Summary
> >       function: Enrichment
> >       privilege: view
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @jjjasper
> 
> The problem in this case, is that which permission is required for each payment type is configured in the service. It is not defined on spec level.

Is that not a problem with having permissions for that service in the spec not necessarily with the multiple access control itself. @jjjasper

File: boat-scaffold/src/main/templates/boat-docs/styles.mustache (hunk 1 above)
Comment (harryMartin678) [reply boat-scaffold/src/main/templates/boat-docs/styles.mustache]: Done

File: boat-scaffold/src/main/templates/boat-docs/styles.mustache (hunk 2 above)
Comment (harryMartin678) [reply boat-scaffold/src/main/templates/boat-docs/styles.mustache]: Done

**Pull request closed without merging**